contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,340 @@
+    // This field should be left empty if specifying non-namespaced resources.
+    // Any empty list implies every namespace.
+    Namespaces []string
+    // Resources is a list of resource types this rule applies to.
+    // Any empty list implies every resource type.
+    Resources []string
+    // ResourceNames is an optional white list of names that the rule applies to.
+    // Any empty list implies everything.
+    ResourceNames []string
+
+    // NonResourceURLs is a set of partial urls that should be audited.
+    // *s are allowed, but only as the full, final step in the path.
+    // If an action is not a resource API request, then the URL is split on '/' and
+    // is checked against the NonResourceURLs to look for a match.
+    NonResourceURLs []string
+}
+```
+
+As an example, the administrator may decide that by default requests should be audited with the
+response, except for get and list requests. On top of that, they wish to completely ignore the noisy
+`kube-proxy` endpoint requests. This looks like:
+
+```yaml
+rules:
+    - level: None
+      users: ["system:kube-proxy"]
+      apiGroups: [""] # The core API group
+      resources: ["endpoints"]
+    - level: RequestBody
+      verbs: ["get", "list"]
+    # The default for non-resource URLs
+    - level: Metadata
+      nonResourceURLs: ["*"]
+    # The default for everything else
+    - level: ResponseBody
+```
+
+The policy is checked immediately after authentication in the request handling, and determines how
+the `audit.Event` is formed.
+
+### Filters
+
+In addition to the high-level policy rules, auditing can be controlled at a more fine-grained level
+with `Filters`. Unlike the policy, filters are applied _after_ the `audit.Event` is constructed, but
+before it's passed to the output backend.
+
+TODO: Define how filters work. They should enable dropping sensitive fields from the
+request/response/storage objects.
+
+### Output Backend Interface
+
+```go
+package audit
+
+type OutputBackend interface {
+  // Thread-safe, blocking.
+  Log(e *Event) error
+}
+```
+
+It is the responsibility of the OutputBackend to manage concurrency, but it is acceptable for the
+method to block. Errors will be handled by recording a count in prometheus, and attempting to write
+to the standard debug log.
+
+
+### Apiserver Command Line Flags
+
+```shell
+$ kube-apiserver --audit-output file:path=/var/log/apiserver-audit.log,rotate=1d,max=1024MB,format=json \
+  --audit-policy user=EndUser,kind=ReplicationController:output=diff,exclude=status
+```
+
+## Sensible (not necessarily sequential) Milestones of Implementation
+
+1. Add `audit.Event` and `audit.OutputBackend` and implement [#27087](https://github.com/kubernetes/kubernetes/pull/27087)'s basic auditing using them, using a single global audit Level, up to `ResponseBody`.
+1. Implement the full `audit.Policy` rule specification.
+1. Add deep inspection on the storage level to the old and the new object.
+1. Add filter support.
```

Comment:
When the time comes, I'd like to see an update describing filters before a pull trying to add them.